### PR TITLE
Implement OAuth

### PR DESCRIPTION
**Pull request by Shre-D**
Adds authentication and stores users on a database.

### CODE REVIEW

**Comment by soumitradev** [comment]
From 02d7bcd onwards some of the commits use past tense for their commit messages, please rebase and reword these commits

File: src/config/authClient.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+//initializes and returns the google oauth client
+export async function getClient() {
+  if (!client) {
+    const googleIssuer = await Issuer.discover("https://accounts.google.com");
+
+    client = new googleIssuer.Client({
+      client_id: env.GOOGLE_CLIENT_ID,
+      client_secret: env.GOOGLE_CLIENT_SECRET,
+      redirect_uris: ["http://localhost:3000/auth/callback"],
```
Comment (soumitradev) [inline src/config/authClient.ts:14]: Use `env.PROD_URL` here, and replace your env prod url with `http://localhost:3000`

File: src/controllers/user/auth.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+import { Session, UserData } from "../../types/auth";
+
+// openid code_verifier
+const code_verifier = generators.codeVerifier();
+
+// redirects to the redirect URL for signing in
+export async function manageAuthRedirect(req: Request, res: Response) {
+  try {
+    if (req.cookies["session"]) {
+      res.redirect("http://localhost:3000/auth/callback");
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:16]: Again, use `env.PROD_URL` here

File: src/controllers/user/auth.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        authenticated: true,
+        message: "user is logged in",
+      });
+    } else {
+      //sets session cookie
+
+      const client = await getClient();
+      const params = client.callbackParams(req);
+
+      //tokenSet contains the refresh_token and access_token codes
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:50]: Again, the code comments not having spaces thing

File: src/controllers/user/auth.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+      const tokenSet = await client.callback(
+        "http://localhost:3000/auth/callback",
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:52]: Again, use `env.PROD_URL` here

File: src/controllers/user/auth.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        params,
+        { code_verifier }
+      );
+
+      //obtaining the acces_token from tokenSet
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:57]: typo + space between comment and slash

File: src/controllers/user/auth.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
+import { TokenSet } from "openid-client";
+import { generators } from "openid-client";
+import { Request, Response } from "express";
+import { getClient } from "../../config/authClient";
+import { userRepository } from "../../repositories/userRepository";
+import { DegreeEnum } from "../../types/degrees";
+import { Session, UserData } from "../../types/auth";
+
+// openid code_verifier
+const code_verifier = generators.codeVerifier();
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:10]: use camelCase for vars

File: src/controllers/user/auth.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+// redirects to the redirect URL for signing in
+export async function manageAuthRedirect(req: Request, res: Response) {
+  try {
+    if (req.cookies["session"]) {
+      res.redirect("http://localhost:3000/auth/callback");
+    } else {
+      const client = await getClient();
+      const code_challenge = generators.codeChallenge(code_verifier);
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:19]: use camelCase for vars

File: src/controllers/user/auth.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+      const auth_redirect = client.authorizationUrl({
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:21]: use camelCase for vars

File: src/controllers/user/auth.ts [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+      const tokenSet = await client.callback(
+        "http://localhost:3000/auth/callback",
+        params,
+        { code_verifier }
+      );
+
+      //obtaining the acces_token from tokenSet
+      const access_token = tokenSet.access_token;
+
+      //obtaining userinfo from the access_token code
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:60]: space between comment and slash

File: src/controllers/user/auth.ts [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+      const userinfo = await client.userinfo(access_token as string | TokenSet);
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:61]: use camelCase for vars

File: src/controllers/user/auth.ts [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        name: userinfo.name,
+        email: userinfo.email,
+      };
+
+      console.log(userData);
+
+      // tokenSet.claims() returns validated information contained upon accessing the token
+      const tokenExpiryTime = tokenSet.claims().exp;
+
+      //defines maxAge to be the time when the session cookie expires
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:73]: space between comment and slash

File: src/controllers/user/auth.ts [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+      const maxAge = tokenExpiryTime * 1000 - Date.now(); //converts into milliseconds
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:74]: space between comment and slash

File: src/controllers/user/auth.ts [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+      //setting the cookie
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:76]: space between comment and slash

File: src/controllers/user/auth.ts [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+  redirect them to a /profile route where they fill their degrees
+  on the frontend
+  */
+    // res.redirect("http://localhost:3000");
+    res.status(401).redirect("http://localhost:3000/auth/login");
+  }
+}
+
+export async function getDegrees(req: Request, res: Response) {
+  /*this function is declared outside the try
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:96]: Don't use multiline comments, use multiple one-line comments

File: src/controllers/user/auth.ts [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+      return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
+    });
+  }
+
+  try {
+    /* for user to enter their degrees */
+    /* the session cookie is parsed here, and then the information about the user,
+  name, email and degrees is stored on the database */
+
+    //gets userinfo as part of session
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:114]: space between comment and slash

File: src/controllers/user/auth.ts [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+    const degrees: DegreeEnum[] = req.body.degrees;
+
+    const userData: UserData = {
+      name: session.name,
+      email: session.email,
+      degrees: degrees,
+    };
+
+    // console.log(userData);
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:126]: remove this comment

File: src/controllers/user/auth.ts [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+    //slices mail to obtain batch
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:128]: space between comment and slash

File: src/controllers/user/auth.ts [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+    const batch = userData.email?.slice(3, 5)!;
+
+    //converts name to title case
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:131]: space between comment and slash

File: src/controllers/user/auth.ts [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+    const name = toTitleCase(userData.name);
+
+    /*
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:134]: Don't use multiline comments, use multiple one-line comments

File: src/controllers/user/auth.ts [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+    this can be better handled by first accessing the find user endpoint,
+    if user exists, then send a response and return. For now, this can just insert.
+    Errors are not handled here, as of yet.
+    */
+    userRepository.insert({
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:139]: Migrate this to query builder pattern, we are migrating the repo to it

File: src/controllers/user/auth.ts [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+    });
+  } catch (error: any) {
+    res.status(500).json({
+      success: false,
+      message: "failed to register",
+    });
+  }
+}
+
+//this function deletes the session cookie to log out
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:158]: space between comment and slash

File: src/controllers/user/auth.ts [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+export async function logout(req: Request, res: Response) {
+  res.clearCookie("session");
+  res.json({
+    authenticated: false,
+    message: "user logged out",
+  });
+}
+
+/* On any route, when checking if a user is logged in, check for the cookie
```
Comment (soumitradev) [inline src/controllers/user/auth.ts:167]: Don't use multiline comments, use multiple one-line comments, also I don't think you should have explanatory comments at the bottom of the file. If you really need these comments, move it to the top of the file

**Review by soumitradev — CHANGES_REQUESTED** [review]
See comments

**Review by soumitradev — APPROVED** [review]
There's still two snake case variables in here, but I'll fix it while integrating auth with the rest of the endpoints. Good job, LGTM, merging.

**Comment by megz15** [comment]
> There's still two snake case variables in here, but I'll fix it while integrating auth with the rest of the endpoints. Good job, LGTM, merging.

Changing those vars to camelCase breaks node-openid-client for some reason

**Pull request merged**